### ISSUE TITLE
Request for help: typed Zod combinator with dynamic field name

### ISSUE THREAD

**Issue by lilred** [vriad/zod#211]
I apologize, this is almost certainly not the correct place to ask, but I'm not sure where else.

I asked [this question](https://stackoverflow.com/questions/64665779/typed-zod-combinator-with-dynamic-key) on Stack Overflow, reproducing it below.

___

My XML to JSON library emits `{MyKey: T}` for one-element lists, and `{MyKey: T[]}` for multi-element lists. The corresponding TypeScript type is `type XmlJsonArray<T, element extends string> = Record<element, T | T[]>`. I've used the following to implement it as a Zod schema:

```typescript
const XmlJsonArray = <T, element extends string>(element: element, schema: z.Schema<T>) => {
  // TODO what is the idiomatic approach here?
  const outerSchema: Record<element, z.Schema<T | T[]>> = {} as any;
  outerSchema[element] = z.union([schema, z.array(schema)]);
  return z.object(outerSchema);
};
```

Is there a way to do this without using `any`?

**Comment by vriad** [comment]
Good question. I actually couldn't figure this out either. Trying to write generic methods that accept Zod schemas is tricky. 

I ended up implementing a new method `.setKey` to allow for this use case. So you'll have to upgrade to `zod@2.0.0-beta.17` or higher (`yarn add zod@beta`) will upgrade you to the newest version.

Here's how it's done:

```ts
const XmlJsonArray = <Key extends string, Schema extends z.Schema<any>>(
  element: Key,
  schema: Schema,
) => {
  return z.object({}).setKey(element, z.union([schema, z.array(schema)]));
};

const test = XmlJsonArray('asdf', z.string());
```

Parsing works as expected:

```ts
// both work
test.parse({ asdf: 'hello' });
test.parse({ asdf: ['hello'] });
```

And type inference works too:

<img width="311" alt="Screen Shot 2020-11-03 at 3 51 42 PM" src="https://user-images.githubusercontent.com/3084745/98039295-7c4f6f00-1dec-11eb-931e-6262fd08e1c3.png">

Good question!

**Comment by lilred** [comment]
This is brilliant, thank you for your help!

**Comment by lilred** [comment]
Follow-up question: is there a reasonably concise way of representing the return type of `XmlJsonArray`?

```typescript
const XmlJsonArray = <tag extends string, T>(
  tag: tag,
  schema: z.ZodType<T>
): z.ZodType<{[t in tag]: T | T[]}> => { // doesn't work
  return z.object({}).setKey(tag, z.union([schema, z.array(schema)]));
};
```

I suspect that I'm running into a hard limitation of TypeScript here, but you'd know better than I.

**Comment by colinhacks** [comment]
You aren't! Just gotta trust the type inference:

```ts
import * as z from '.'

const XmlJsonArray = <tag extends string, T>(
  tag: tag,
  schema: z.ZodType<T>,
) => {
  return z.object({}).setKey(tag, z.union([schema, z.array(schema)]));
};

const mySchema = XmlJsonArray('tuna', z.string());
type mySchema  = z.infer<typeof mySchema>
// => { tuna: string | string[]; }
```

In general things get funky when you use `z.ZodType` since it's an abstract base class that isn't really intended for actual use. At some point I need to write up a whole guide about building generic functions on top of Zod that walk through some of these best practices.

**Comment by lilred** [comment]
>So you want to explicitly annotate the return type of the function? Any particular reason?

I'm trying to feed the parsed output of `XmlJsonArray` to a function that converts it into a good old JavaScript array. I'm having some weird type mismatches that I'm trying to work out. From typing up the explanation below I think my problem is that `setKey` creates an optional field.

```typescript
import * as z from 'zod';

const XmlJsonArray = <tag extends string, T, U extends z.Schema<T>>(tag: tag, schema: U) => {
  return z.object({}).setKey(tag, z.union([schema, z.array(schema)]));
  // or this syntax I just learned about
  return z.object({[tag]: z.union([schema, z.array(schema)])});
};

// TODO the field in a should NOT be optional!
const formatXmlJsonArray = <tag extends string, T>(tag: tag, a: {[v in tag]?: T | T[]}): T[] => {
  const elem: T | T[] = a[tag];
  if (elem instanceof Array) return elem;
  return [elem];
};

// Example usage

const numberParser = z.number();

// const zeroElements = '\r\n';
const oneElement = {Number: 1};
const manyElements = {Number: [1, 2, 3]};

// formatNumberXmlJsonArray('Number', NumberXmlJsonArray.parse(zeroElements)); // should return []
formatXmlJsonArray('Number', XmlJsonArray('Number', numberParser).parse(oneElement)); // should return [1]
formatXmlJsonArray('Number', XmlJsonArray('Number', numberParser).parse(manyElements)); // should return [1, 2, 3]
```

TL;DR two things of interest:

- The `{[dynamicFieldName]: fieldValue}` ES6 syntax for creating objects with dynamic field names
- The fact that setKey appears to create optional fields (by design? Bug?)

**Comment by colinhacks** [comment]
I'm having trouble duplicating the `setKey` creating optional fields error. Make sure you have `strictNullChecks: true` in your tsconfig `"compilerOptions"`.

![Screen Shot 2020-11-28 at 2 20 48 PM](https://user-images.githubusercontent.com/3084745/100527204-0806b200-3185-11eb-9033-53e00d6b9e5f.png)

I got this working. You need to use conditional types to get `formatXmlJsonArray` to have the output type signature vary based on the type of the input (which is what you were trying to achieve). Dive into the docs here to learn more:


```ts
import * as z from '.';

const XmlJsonArray = <tag extends string, T, U extends z.Schema<T>>(
  tag: tag,
  schema: U,
) => {
  return z.object({}).setKey(tag, z.union([schema, z.array(schema)]));
  // or this syntax I just learned about
  // return z.object({ [tag]: z.union([schema, z.array(schema)]) });
};

// TODO the field in a should NOT be optional!
const formatXmlJsonArray = <tag extends string, T>(
  tag: tag,
  a: { [v in tag]?: T },
): T extends Array<any> ? T : T[] => {
  const elem = a[tag] as any;
  if (elem instanceof Array) return elem as any;
  return [elem] as any;
};

// Example usage

const numberParser = z.number();

// const zeroElements = '\r\n';
const oneElement = { Number: 1 };
const manyElements = { Number: [1, 2, 3] };

// formatXmlJsonArray('Number', XmlJsonArray('Number', numberParser).parse(zeroElements)); // should return []
const sample1 = formatXmlJsonArray(
  'Number',
  XmlJsonArray('Number', numberParser).parse(oneElement),
); // should return [1]

const sample2 = formatXmlJsonArray(
  'Number',
  XmlJsonArray('Number', numberParser).parse(manyElements),
); // should return [1, 2, 3]
```

**Comment by lilred** [comment]
>Make sure you have `"strictNullChecks": true` (or just `"strict": true`) in your tsconfig `"compilerOptions"` — you'll see strange behavior unless that setting is set.

Yep, that was it. Thank you so much for your time!

**Comment by lilred** [comment]
I apologize if this is silly, but just in case you missed it:

>* The `{[dynamicFieldName]: fieldValue}` ES6 syntax for creating objects with dynamic field names

I suspect that this makes `setKey` redundant.

**Comment by colinhacks** [comment]
I was aware of this syntax but I couldn't get it to work regardless. Consider this implementation:

```ts
const test = <T extends string>(key:T)=>{
  return {[key]: 25};
}
```

TypeScript isn't able to infer the return type (`{ asdf: number}`) using the dynamic field names syntax. 

![Screen Shot 2020-12-01 at 10 21 11 PM](https://user-images.githubusercontent.com/3084745/100836143-a3b54e00-3423-11eb-80fa-207ee9eed8ad.png)

**Comment by lilred** [comment]
[I opened an issue on the TypeScript repo.](https://github.com/microsoft/TypeScript/issues/41779) Cheers!

**Comment by nemonemi** [comment]
Could you please document the `setKey`?
There is no mention of it in the documentation at the moment.

**Comment by scotttrinh** [comment]
@nemonemi 

The problem you're running into is that `setKey` does not mutate the schema. It returns a new schema, so you need to reassign your schema variable. This works:

[CodeSandbox](https://codesandbox.io/s/focused-pine-m7xfq?file=/src/index.ts)
```ts
const someData = [{ name: "foo" }, { name: "bar" }];

let schema = z.object({});
for (const prop of someData) {
  schema = schema.setKey(prop.name, z.string());
}
```

**Comment by nemonemi** [comment]
Love it!
Thanks, @scotttrinh!
Why doesn't the documentation convey anything about the `setKey` method?

**Comment by scotttrinh** [comment]
No particular reason other than we just didn't write anything about it! I'll write something up later today, thanks for the encouragement.

**Comment by CSFlorin** [comment]
@scotttrinh I just looked but couldn't find anything about setKey in the docs. I don't see a JSDoc comment either. How's the documentation coming along?

**Comment by k-kyler** [comment]
> @nemonemi
> 
> The problem you're running into is that `setKey` does not mutate the schema. It returns a new schema, so you need to reassign your schema variable. This works:
> 
> [CodeSandbox](https://codesandbox.io/s/focused-pine-m7xfq?file=/src/index.ts)
> 
> ```ts
> const someData = [{ name: "foo" }, { name: "bar" }];
> 
> let schema = z.object({});
> for (const prop of someData) {
>   schema = schema.setKey(prop.name, z.string());
> }
> ```
> 
> or if you prefer to avoid variable mutation:
> 
> ```ts
> const schema = someData.reduce(
>   (schema, definition) => schema.setKey(definition.name, z.string()),
>   z.object({})
> );
> ```

I was looking around for a solution to loop through and create a dynamic schema.

Thank to your solution @scotttrinh, I have solved my problem :D
